graph/pkg/service/v0/drives.go

```diff
@@ -40,57 +40,91 @@ func (g Graph) GetDrives(w http.ResponseWriter, r *http.Request) {
 		errorcode.InvalidRequest.Render(w, r, http.StatusBadRequest, err.Error())
 		return
 	}
-	g.logger.Info().Msg("Calling GetDrives")
+	g.logger.Info().Interface("query", r.URL.Query()).Msg("Calling GetDrives")
 	ctx := r.Context()
 
-	client, err := g.GetClient()
+	filters, err := generateCs3Filters(odataReq)
 	if err != nil {
-		g.logger.Err(err).Msg("error getting grpc client")
+		g.logger.Err(err).Interface("query", r.URL.Query()).Msg("query error")
+		errorcode.NotSupported.Render(w, r, http.StatusNotImplemented, err.Error())
+		return
+	}
+	res, err := g.ListStorageSpacesWithFilters(ctx, filters)
+	switch {
+	case err != nil:
+		g.logger.Error().Err(err).Msg("error sending list storage spaces grpc request")
 		errorcode.GeneralException.Render(w, r, http.StatusInternalServerError, err.Error())
 		return
+	case res.Status.Code != cs3rpc.Code_CODE_OK:
+		if res.Status.Code == cs3rpc.Code_CODE_NOT_FOUND {
+			// return an empty list
+			render.Status(r, http.StatusOK)
+			render.JSON(w, r, &listResponse{})
+			return
+		}
+		g.logger.Error().Err(err).Msg("error sending list storage spaces grpc request")
+		errorcode.GeneralException.Render(w, r, http.StatusInternalServerError, res.Status.Message)
+		return
 	}
 
-	permissions := make(map[string]struct{}, 1)
-	s := sproto.NewPermissionService("com.owncloud.api.settings", grpc.DefaultClient)
+	wdu, err := url.Parse(g.config.Spaces.WebDavBase + g.config.Spaces.WebDavPath)
+	if err != nil {
+		g.logger.Error().Err(err).Msg("error parsing url")
+		errorcode.GeneralException.Render(w, r, http.StatusInternalServerError, err.Error())
+		return
+	}
+	files, err := g.formatDrives(ctx, wdu, res.StorageSpaces)
+	if err != nil {
+		g.logger.Error().Err(err).Msg("error encoding response as json")
+		errorcode.GeneralException.Render(w, r, http.StatusInternalServerError, err.Error())
+		return
+	}
 
-	_, err = s.GetPermissionByID(ctx, &sproto.GetPermissionByIDRequest{
-		PermissionId: settingsSvc.ListAllSpacesPermissionID,
-	})
+	render.Status(r, http.StatusOK)
+	render.JSON(w, r, &listResponse{Value: files})
+}
 
-	// No error means the user has the permission
-	if err == nil {
-		permissions[settingsSvc.ListAllSpacesPermissionName] = struct{}{}
+// GetSingleDrive does a lookup of a single space by spaceId
+func (g Graph) GetSingleDrive(w http.ResponseWriter, r *http.Request) {
+	sanitizedPath := strings.TrimPrefix(r.URL.Path, "/graph/v1.0/")
+	driveId := chi.URLParam(r, "driveID")
+	if driveId == "" {
+		err := fmt.Errorf("no valid space id retrieved")
+		g.logger.Err(err)
+		errorcode.GeneralException.Render(w, r, http.StatusInternalServerError, err.Error())
+		return
 	}
-	value, err := json.Marshal(permissions)
+	// Add the driveId filter to the request query
+	spaceFilterValue := fmt.Sprintf("id eq %s", driveId)
+	query := r.URL.Query()
+	query.Add("$filter", spaceFilterValue)
+	// Parse the request with odata parser
+	odataReq, err := godata.ParseRequest(r.Context(), sanitizedPath, query)
```

Comment:
I think it would be better to create the filter directly (or write a function which takes the id and returns the filter) instead of creating the odataRequest and then transforming it.

Reply:
hmm. I thought we should use the odata parser and route all filters trough it because we will have combinations of filters in the future. The odata parser will then evaluate the filter syntax for us. IMO that parser can do that for us and we should leverage it.

Reply:
adding and modifying filters on different layers at the same time would make the implementation hard to test also.

Reply:
Using the godata parser was my intention back in the days, since it is [well tested](https://github.com/crestonbunch/godata/blob/303b6c35d5bfd48f9f98761a9ec8db87d54c6caa/filter_parser_test.go). Just understand the cases a filter is malformed and how the library behaves.

Reply:
I get that but in this case you are creating the odata query yourself in line 98. That means you already know what filter you expect and because of that could directly call the function `createSpaceIdFilter` which could also be used by `generateCs3Filters`. This would make it actually simpler to test since you've got smaller units.

Reply:
yeah, true, but if we want to adopt the godata parser we might as well use it 🤷. I see it as a "sanitizer" of sorts. In this case this logic can probably be cleaned up a bit. Also now I take another read at it it is a bit fishy that we're appending `id eq ${driveID}` to the request itself.

Seeing this holistically there is no need for an odataRequest, since the driveID is already in the url:

```
r.Route("/{driveID}", func(r chi.Router) {
						r.Patch("/", svc.UpdateDrive)
						r.Get("/", svc.GetSingleDrive)
					})
```

in `{driveID}`. I would use the parser when there is actually a query string in the url, and .... parse it. But for this endpoint I think it is not necessary and @C0rby is right.

Reply:
Also right. BUT

We will have odata queries on that endpoint too like $select and $expand. My idea was to create the full query before we hand it over to the odata parser because we cannot assume that this filter is the only possible query part.

Reply:
We will have a $select for the etag very soon. Because returning the etag would require to do an additional stat request. Therefore we want to let the client decide if he needs it or not.

Reply:
but having a filter on id and an endpoint is redundant and prone to errors! you can get the case the query is different than the endpoint id. Not here because you're hard-coding it, but you're opening the floodgates.

Reply:
I think you misunderstood me. You cannot predict how users will use filters and queries on that endpoint.

Reply:
let me put an example:

`GET [...]/graph/v1.0/drives/1059deca-7917-11ec-88a2-6fd277313577?$filter=id eq '821e8880-7917-11ec-bddd-530fc783a68f'`

Which query is more valid there? As per your code, the `{driveID}` and not the query string. My argument is that there is no need for filtering here at all, since you're only requesting a single resource. Filters are required when you, well, want to limit the result set.

Reply:
https://www.odata.org/documentation/odata-version-3-0/url-conventions/

section 4.1 of that document should shed some light. See how there are no filters on single entities, because it doesn't make sense. You don't have a set to reduce, you either have one, or not.

Reply:
ok, see my updated code. I create the filter directly. select and expand will use the odata parser.